### PR TITLE
Adding Netty Metric to frontend

### PR DESCRIPTION
**Pull request by justinlin-linkedin**
Adding a NettyMetrics class to collect netty buffer related metrics.

### CODE REVIEW

File: ambry-commons/src/main/java/com.github.ambry.commons/NettyMetrics.java [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+        numHeapTotalDeallocations = heapArenaMetrics.stream().mapToLong(PoolArenaMetric::numDeallocations).sum();
+        numHeapTotalActiveAllocations =
+            heapArenaMetrics.stream().mapToLong(PoolArenaMetric::numActiveAllocations).sum();
+        numDirectTotalAllocations = directArenaMetrics.stream().mapToLong(PoolArenaMetric::numAllocations).sum();
+        numDirectTotalDeallocations = directArenaMetrics.stream().mapToLong(PoolArenaMetric::numDeallocations).sum();
+        numDirectTotalActiveAllocations =
+            directArenaMetrics.stream().mapToLong(PoolArenaMetric::numActiveAllocations).sum();
+
+        try {
+          Thread.sleep(config.nettyMetricsRefreshIntervalSeconds * 1000);
```
Comment (jsjtzyy) [inline ambry-commons/src/main/java/com.github.ambry.commons/NettyMetrics.java]: Can we use a `ScheduledExecutorService` to do this? Feel free to refer to `HistogramDumper` in `NonBlockingRouterMetrics`

Comment (zzmao) [reply ambry-commons/src/main/java/com.github.ambry.commons/NettyMetrics.java]: +1

Comment (justinlin-linkedin) [reply ambry-commons/src/main/java/com.github.ambry.commons/NettyMetrics.java]: updated

File: ambry-commons/src/main/java/com.github.ambry.commons/NettyMetrics.java [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+    return numDirectTotalDeallocations;
+  }
+
+  private long getNumDirectTotalActiveAllocations() {
+    return numDirectTotalActiveAllocations;
+  }
+
+  private void register() {
+    registry.register(MetricRegistry.name(NettyMetrics.class, "NumberDirectArenas"),
+        (Gauge<Integer>) () -> getNumDirectArenas());
```
Comment (jsjtzyy) [inline ambry-commons/src/main/java/com.github.ambry.commons/NettyMetrics.java]: optional: change following to method reference, just you did in run method : `PoolArenaMetric::numAllocations`

File: ambry-commons/src/main/java/com.github.ambry.commons/NettyMetrics.java [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+        logger.trace("De-scheduled for collecting netty metrics");
+      }
+    }
+  }
+
+  private class NettyMetricCollector implements Runnable {
+
+    @Override
+    public void run() {
+      PooledByteBufAllocatorMetric metric = PooledByteBufAllocator.DEFAULT.metric();
```
Comment (jsjtzyy) [inline ambry-commons/src/main/java/com.github.ambry.commons/NettyMetrics.java]: please ensure when the method is first called, the metric is not null. If the config is off (netty bytebuf is not enabled) should we start NettyMetrics?

File: ambry-commons/src/main/java/com.github.ambry.commons/NettyMetrics.java (hunk 2 above)
Comment (justinlin-linkedin) [reply ambry-commons/src/main/java/com.github.ambry.commons/NettyMetrics.java]: Gauge is not a functional interface, so I can't use NettyMetrics::getXXXX.

File: ambry-commons/src/main/java/com.github.ambry.commons/NettyMetrics.java (hunk 3 above)
Comment (justinlin-linkedin) [reply ambry-commons/src/main/java/com.github.ambry.commons/NettyMetrics.java]: PooledByteBuf would create metric even if we don't use it. As long as we import this library, the metric will be created. I will not start NettyMetric if we don't enable it in Network layer.

File: ambry-commons/src/main/java/com.github.ambry.commons/NettyMetrics.java (hunk 2 above)
Comment (jsjtzyy) [reply ambry-commons/src/main/java/com.github.ambry.commons/NettyMetrics.java]: How about this? 
```
registry.register(MetricRegistry.name(NettyMetrics.class, "NumberDirectArenas"),
        (Gauge<Integer>) this::getNumDirectArenas);
```

Comment (justinlin-linkedin) [reply ambry-commons/src/main/java/com.github.ambry.commons/NettyMetrics.java]: it worked, updated.

File: ambry-commons/src/main/java/com.github.ambry.commons/NettyMetrics.java [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+ * To Stop collecting, please do
+ * <p>
+ *   nettyMetrics.stop();
+ * </p>
+ */
+public class NettyMetrics {
+
+  private static final Logger logger = LoggerFactory.getLogger(NettyMetrics.class);
+  private volatile MetricRegistry registry;
+  private volatile NettyConfig config;
```
Comment (jsjtzyy) [inline ambry-commons/src/main/java/com.github.ambry.commons/NettyMetrics.java]: why `registry` and `config` are volatile?

**Pull request merged**